tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,313 @@
+use super::Semaphore;
+use crate::loom::cell::UnsafeCell;
+use std::error::Error;
+use std::fmt;
+use std::future::Future;
+use std::mem::MaybeUninit;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A thread-safe cell which can be written to only once.
+///
+/// Provides the functionality to either set the value, in case `OnceCell`
+/// is uninitialized, or get the already initialized value by using an async
+/// function via [`OnceCell::get_or_init_with`] or by using a Future via
+/// [`OnceCell::get_or_init`] directly via [`OnceCell::get_or_init`].
+///
+/// [`OnceCell::get_or_init_with`]: crate::sync::OnceCell::get_or_init_with
+/// [`OnceCell::get_or_init`]: crate::sync::OnceCell::get_or_init
+///
+/// # Examples
+/// ```
+/// use tokio::sync::OnceCell;
+///
+/// async fn some_computation() -> u32 {
+///     1 + 1
+/// }
+///
+/// static ONCE: OnceCell<u32> = OnceCell::new();
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let result1 = tokio::spawn(async {
+///         ONCE.get_or_init_with(some_computation).await
+///     }).await.unwrap();
+///     assert_eq!(*result1, 2);
+/// }
+/// ```
+pub struct OnceCell<T> {
+    value_set: AtomicBool,
+    value: UnsafeCell<MaybeUninit<T>>,
+    semaphore: Semaphore,
+}
+
+impl<T> Default for OnceCell<T> {
+    fn default() -> OnceCell<T> {
+        OnceCell::new()
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for OnceCell<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("OnceCell")
+            .field("value", &self.get())
+            .finish()
+    }
+}
+
+impl<T: Clone> Clone for OnceCell<T> {
+    fn clone(&self) -> OnceCell<T> {
+        let new_cell = OnceCell::new();
+        if let Ok(value) = self.get() {
+            match new_cell.set(value.clone()) {
+                Ok(()) => (),
+                Err(_) => unreachable!(),
+            }
+        }
+        new_cell
```

Comment:
It would be possible to provide an `set_mut` method that takes an `&mut self` and always sets the value, overwriting the previous value. This would be safe because it having mutable access means that nobody is accessing the `OnceCell` concurrently.

Another argument for it being safe is that you can already implement it now with a safe API.
```Rust
let new_cell = OnceCell::new();
new_cell.set(new_value).unwrap();
*the_cell = new_cell;
```

Reply:
Doesn't this contradict the use case for `OnceCell` (a cell which can only be written to once)?

Reply:
You don't have to remove any functionality to add this feature.

Reply:
Yes, but I'm just wondering whether this is something we want to have, given that `OnceCell` by definition should only be able to be set once.

Reply:
That's a bit of a philosophical discussion. I don't think there's anything wrong with it. What you can do with `&mut OnceCell` is not the same as what you can do with a `&OnceCell`.